crates/stages/src/stages/sender_recovery.rs

```diff
@@ -91,24 +77,54 @@ impl<DB: Database> Stage<DB> for SenderRecoveryStage {
 
         // Iterate over transactions in chunks
         info!(target: "sync::stages::sender_recovery", start_tx_index, end_tx_index, "Recovering senders");
-        for chunk in &entries.chunks(self.batch_size) {
-            let transactions = chunk.collect::<Result<Vec<_>, DbError>>()?;
-            // Recover signers for the chunk in parallel
-            let recovered = transactions
-                .into_par_iter()
-                .map(|(tx_id, transaction)| {
-                    trace!(target: "sync::stages::sender_recovery", tx_id, hash = ?transaction.hash(), "Recovering sender");
-                    let signer =
-                        transaction.recover_signer().ok_or_else::<StageError, _>(|| {
-                            SenderRecoveryStageError::SenderRecovery { tx: tx_id }.into()
-                        })?;
+
+        // a channel to receive results from a rayon job
+        let (tx, rx) = mpsc::unbounded_channel();
+
+        // spawn recovery jobs onto the default rayon threadpool and send the result through the
+        // channel
+        for entry in entries {
+            let (tx_id, transaction) = entry?;
+            let tx = tx.clone();
+            rayon::spawn(move || {
+                trace!(target: "sync::stages::sender_recovery", tx_id, hash = ?transaction.hash(), "Recovering sender");
+                let res = if let Some(signer) = transaction.recover_signer() {
                     Ok((tx_id, signer))
-                })
-                .collect::<Result<Vec<_>, StageError>>()?;
-            // Append the signers to the table
-            recovered.into_iter().try_for_each(|(id, sender)| senders_cursor.append(id, sender))?;
+                } else {
+                    Err(StageError::from(SenderRecoveryStageError::SenderRecovery { tx: tx_id }))
+                };
+                // send the result back
+                let _ = tx.send(res);
+            });
         }
 
+        drop(tx);
+
+        let mut recovered_senders = UnboundedReceiverStream::new(rx);
+
+        while let Some(recovered) = recovered_senders.next().await {
+            let (id, sender) = recovered?;
+            senders_cursor.append(id, sender)?;
```

Comment:
can't see anything that would guarantee that we receive transaction senders in a preserved order. the `append` will fail otherwise

Reply:
this might be the reason why tests are failing

Reply:
ah I see, so `into_par_iter` actually maintains order?
can we use `cursor.insert` instead?

Reply:
ah okay, the `rayon.collect` stores the results in order TIL.

we can do something similar since we know the order

Reply:
just confirming that this is the problem

```
---- stages::sender_recovery::tests::execute_sender_recovery stdout ----
[crates/stages/src/stages/sender_recovery.rs:107] &id = 2
[crates/stages/src/stages/sender_recovery.rs:107] &sender = 0xd1438bab5f4ccf2a8db19e4eafdb654ab701d7e4
[crates/stages/src/stages/sender_recovery.rs:107] &id = 9
[crates/stages/src/stages/sender_recovery.rs:107] &sender = 0x7e4eed7556609301aad659fdf793e7505dd4e457
[crates/stages/src/stages/sender_recovery.rs:107] &id = 5
[crates/stages/src/stages/sender_recovery.rs:107] &sender = 0xc205b84d0fd77ab5a835bcb06256c3dea2bcabe0
thread 'stages::sender_recovery::tests::execute_sender_recovery' panicked at 'assertion failed: `Err(Database(Write(4294936878)))` does not match `Ok(ExecOutput { done, stage_progress }) if
done && stage_progress == previous_stage`', crates/stages/src/stages/sender_recovery.rs:155:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

`insert` won't work, it's too slow

Reply:
added a stream type that emits in sequence

this branch

```
Stages/SendersRecovery-batch-1000
                        time:   [14.001 s 15.219 s 16.241 s]

```

main

```
Stages/SendersRecovery-batch-1000
                        time:   [17.191 s 17.890 s 18.557 s]

```

Reply:
made some additional changes to the stream type and now it's down to 

```
Stages/SendersRecovery-batch-1000
                        time:   [10.075 s 10.635 s 11.220 s]

```